### PR TITLE
[RFC] Preserve new lines next to a + where either side is a string

### LINKED ISSUE #489: Put a hardline between "StringLiteral" + "StringLiteral"
Right now

```js
    console.error(
      'In order to use ' + prompt + ', you need to configure a ' +
      'few environment variables to be able to commit to the ' +
      'repository. Follow those steps to get you setup:\n' +
      '\n' +
      'Go to https://github.com/settings/tokens/new\n' +
      ' - Fill "Token description" with "' + prompt + ' for ' +
        repoSlug + '"\n' +
      ' - Check "public_repo"\n' +
      ' - Press "Generate Token"\n' +
      '\n' +
      'In a different tab, go to https://travis-ci.org/' +
        repoSlug + '/settings\n' +
      ' - Make sure "Build only if .travis.yml is present" is ON\n' +
      ' - Fill "Name" with "GITHUB_USER" and "Value" with the name of the ' +
        'account you generated the token with. Press "Add"\n' +
      '\n' +
      'Once this is done, commit anything to the repository to restart ' +
        'Travis and it should work :)'
    );
```

gets printed as

```js
    console.error(
      "In order to use " +
        prompt +
        ", you need to configure a " +
        "few environment variables to be able to commit to the " +
        "repository. Follow those steps to get you setup:\n" +
        "\n" +
        "Go to https://github.com/settings/tokens/new\n" +
        ' - Fill "Token description" with "' +
        prompt +
        " for " +
        repoSlug +
        '"\n' +
        ' - Check "public_repo"\n' +
        ' - Press "Generate Token"\n' +
        "\n" +
        "In a different tab, go to https://travis-ci.org/" +
        repoSlug +
        "/settings\n" +
        ' - Make sure "Build only if .travis.yml is present" is ON\n' +
        ' - Fill "Name" with "GITHUB_USER" and "Value" with the name of the ' +
        'account you generated the token with. Press "Add"\n' +
        "\n" +
        "Once this is done, commit anything to the repository to restart " +
        "Travis and it should work :)"
    );
```

which breaks on every single + which makes the code harder to read.

One pattern that we can extract is that if you have "string" + "string", it likely means that there was a \n there originally. So, if we would reformat the code as

```js
group("In order to use " + prompt + ", you need to configure a " +)
group("few environment variables to be able to commit to the " +)
group("repository. Follow those steps to get you setup:\n" +)
group("\n" +)
group("Go to https://github.com/settings/tokens/new\n" +)
group(' - Fill "Token description" with "' + prompt + " for " + repoSlug + '"\n' +)
group(' - Check "public_repo"\n' +)
group(' - Press "Generate Token"\n' +)
group("\n" +)
group("In a different tab, go to https://travis-ci.org/" + repoSlug + "/settings\n" +)
group(' - Make sure "Build only if .travis.yml is present" is ON\n' +)
group(' - Fill "Name" with "GITHUB_USER" and "Value" with the name of the ' +)
group('account you generated the token with. Press "Add"\n' +)
group("\n" +)
group("Once this is done, commit anything to the repository to restart " +)
group("Travis and it should work :)")
```

that would lead to a better output than breaking on every single +

### PR DESCRIPTION
**Pull request by vjeux**
Many people still use string concatenation in order to build a bigger one and new lines are usually significant. This is an attempt at solving this use case, right now we split on every single `+` which is making the string lose meaning compared to the original one.

Fixes #489

### CODE REVIEW

File: src/printer.js [hunk 1]
```diff
@@ -1117,10 +1117,7 @@ function genericPrintNoParens(path, options, print) {
 
       if (n.value) {
         let res;
-        if (
-          (n.value.type === "StringLiteral" || n.value.type === "Literal") &&
-            typeof n.value.value === "string"
-        ) {
+        if (isStringLiteral(n.value)) {
```
Comment (vjeux) [inline src/printer.js]: I created a helper function for this.

File: tests/string_concat/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    \'account you generated the token with. Press \"Add\"\\n\' +
+  \'\\n\' +
+  \'Once this is done, commit anything to the repository to restart \' +
+    \'Travis and it should work :)\'
+);
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+var fnString = \'\"\' + this.USE + \" \" + this.STRICT + \'\";\\n\' +
+  this.filterPrefix() +
+  \"var fn=\" + this.generateFunction(\"fn\", \"s,l,a,i\") +
+  extra + this.watchFns() +
```
Comment (vjeux) [inline tests/string_concat/__snapshots__/jsfmt.spec.js.snap]: Notice how `extra` doesn't have a string literal around so it isn't on its own line.

File: tests/string_concat/__snapshots__/jsfmt.spec.js.snap [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    \' - Press \"Generate Token\"\\n\' +
+    \"\\n\" +
+    \"In a different tab, go to https://travis-ci.org/\" +
+    repoSlug + \"/settings\\n\" +
+    \' - Make sure \"Build only if .travis.yml is present\" is ON\\n\' +
+    \' - Fill \"Name\" with \"GITHUB_USER\" and \"Value\" with the name of the \' +
+    \'account you generated the token with. Press \"Add\"\\n\' +
+    \"\\n\" +
+    \"Once this is done, commit anything to the repository to restart \" +
+    \"Travis and it should work :)\"
```
Comment (vjeux) [inline tests/string_concat/__snapshots__/jsfmt.spec.js.snap]: This whole expression has been left untouched :)

**Comment by jlongster** [comment]
The reason I think this won't work is because you still have expressions like this (taken from CRA):

```js
console.log('The directory ' + chalk.green(name) + ' contains files that could conflict.');
```

You'll have a mix of expressions so you can't just check string literals. I think we should close this and move forward with the general escape hatch in https://github.com/jlongster/prettier/pull/671

**Comment by vjeux** [comment]
Will close in favor of the escape hatch.

**Pull request closed without merging**